File: services/order-service/src/main/java/com/bookland/order/entity/Bill.java
package com.bookland.order.entity;

import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "bill")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Bill {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "user_id", nullable = false)
    private String userId; // maps to user's email injected from gateway X-User-Email

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "payment_method_id", nullable = false)
    private PaymentMethod paymentMethod;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "shipping_method_id", nullable = false)
    private ShippingMethod shippingMethod;

    @Column(name = "total_cost", nullable = false)
    private Double totalCost;

    @Column(name = "approved_by")
    private String approvedBy; // approver email / username

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    @Builder.Default
    private BillStatus status = BillStatus.PENDING;

    @Column(name = "payment_status", length = 20)
    @Builder.Default
    private String paymentStatus = "PENDING";

    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @Column(name = "approved_at")
    private LocalDateTime approvedAt;

    @OneToMany(mappedBy = "bill", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    @Builder.Default
    private List<BillBook> billBooks = new ArrayList<>();

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

    public enum BillStatus {
        PENDING, APPROVED, SHIPPING, SHIPPED, COMPLETED, CANCELED
    }
}

File: BookLand_FE/src/types/Bill.ts
import type { BillBook } from './BillBook.ts';
import type { PaymentTransaction } from './PaymentTransaction.ts';
import type { EventLog } from './EventLog.ts';

export const BillStatus = {
    PENDING: 'PENDING',
    APPROVED: 'APPROVED',
    SHIPPING: 'SHIPPING',
    SHIPPED: 'SHIPPED',
    COMPLETED: 'COMPLETED',
    CANCELED: 'CANCELED'
} as const;

export type BillStatus = (typeof BillStatus)[keyof typeof BillStatus];

export interface Bill {
    id: number;
    userId: number;
    userName: string;
    fullName?: string;
    email?: string;
    phone?: string;
    address?: string;
    city?: string;
    notes?: string;
    paymentMethodId: number;
    paymentMethodName: string;
    shippingMethodId: number;
    shippingMethodName: string;
    shippingCost: number;
    totalCost: number;
    approvedById?: number;
    approvedByName?: string;
    status: BillStatus;
    paymentStatus?: string;
    createdAt?: string;
    updatedAt?: string;
    approvedAt?: string;
    books?: BillBook[];
    transactions?: PaymentTransaction[];
    eventLogs?: EventLog[];
}

export interface CreateBillRequest {
    userId: number;
    fullName: string;
    email: string;
    phone: string;
    address: string;
    city: string;
    postalCode?: string;
    notes?: string;
    paymentMethodId: number;
    shippingMethodId: number;
    books: {
        bookId: number;
        quantity: number;
    }[];
}

export interface UpdateBillStatusRequest {
    status: BillStatus;
    note?: string;
}
